cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -0,0 +1,224 @@
+import com.google.common.util.concurrent.AbstractIdleService;
+import io.cdap.cdap.api.security.store.SecureStore;
+import io.cdap.cdap.common.utils.DirUtils;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfig;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfigValidationException;
+import org.eclipse.jgit.api.CloneCommand;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.api.LsRemoteCommand;
+import org.eclipse.jgit.api.TransportCommand;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.eclipse.jgit.api.errors.TransportException;
+import org.eclipse.jgit.internal.storage.file.FileRepository;
+import org.eclipse.jgit.lib.Ref;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Collections;
+import javax.annotation.Nullable;
+
+/**
+ * A git repository manager that is responsible for handling interfacing with git. It fetches credentials
+ * from a {@link io.cdap.cdap.api.security.store.SecureStore}, builds an {@link AuthenticationStrategy} for different
+ * Git hosting services and provides version control operations.
+ */
+public class RepositoryManager extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(RepositoryManager.class);
+  private final SecureStore secureStore;
+  private final SourceControlConfig sourceControlConfig;
+  private Git git;
+
+  public RepositoryManager(SecureStore secureStore, SourceControlConfig sourceControlConfig) {
+    this.secureStore = secureStore;
+    this.sourceControlConfig = sourceControlConfig;
+  }
+
+  /**
+   * Returns the base path in the git repo to store CDAP applications. If an optional Path prefix is provided in the
+   * repository configuration, the returned path includes it.
+   *
+   * @return the path for the repository base directory.
+   */
+  public Path getBasePath() {
+    Path localRepoPath = sourceControlConfig.getLocalRepoPath();
+    String pathPrefix = sourceControlConfig.getRepositoryConfig().getPathPrefix();
+    if (pathPrefix == null) {
+      return localRepoPath;
+    }
+    return localRepoPath.resolve(pathPrefix);
+  }
+
+  /**
+   * Validates the provided configuration.
+   *
+   * @param secureStore         A secure store for fetching credentials if required.
+   * @param sourceControlConfig Configuration for source control operations.
+   * @throws IOException                         when there are File I/O problems.
+   * @throws RepositoryConfigValidationException when provided repository configuration is invalid.
+   */
+  public static void validateConfig(SecureStore secureStore, SourceControlConfig sourceControlConfig) throws
+    IOException, RepositoryConfigValidationException {
+    try {
+      doValidate(secureStore, sourceControlConfig);
+    } finally {
+      deletePathIfExists(sourceControlConfig.getReposBasePath());
```

Comment:
It's a bit strange to have the validate method doing mutation. Why is it so? What actually is the contract of validation?

Reply:
Thanks for the idea about using InMemoryRepo, wasn't aware of that. I've changed to use that, so no cleanup is required now.